src/libYARP_dev/include/yarp/dev/IFrameSource.h

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright (C) 2006-2018 Istituto Italiano di Tecnologia (IIT)
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms of the
+ * BSD-3-Clause license. See the accompanying LICENSE file for details.
+ */
+
+#ifndef YARP_DEV_IFRAMESOURCE_H
+#define YARP_DEV_IFRAMESOURCE_H
+
+#include <unordered_set>
+#include <map>
+#include <vector>
+#include <functional>
+#include <robotology/Result.h>
+#include <yarp/dev/api.h>
+#include <yarp/math/Quaternion.h>
+#include <yarp/math/FrameTransform.h>
+
+namespace yarp {
+    namespace dev {
+        class IFrameSource;
+    }
+}
+
+struct CustomFrameCompare
+{
+    bool operator()(const yarp::math::FrameTransform& lhs, const yarp::math::FrameTransform& rhs)
+    {
+        if (lhs.frameId == rhs.frameId)
+            return true;
+    }
+};
+
+class FrameEditor
+{
+    std::map<std::string, yarp::math::FrameTransform> storage;
+    yarp::math::FrameTransform nullframe;
+public:
+    const auto begin() const { return storage.begin(); }
+    const auto end() const { return storage.end(); }
+    bool insertUpdate(const yarp::math::FrameTransform& frame)
+    {
+        storage[frame.frameId] = frame;
+        return true;
+    }
+
+    robotology::Result<yarp::math::FrameTransform> get(const std::string& frameid)
+    {
+        if (storage.find(frameid) == storage.end())
+        {
+            return { false, 0, "", nullframe };
+        }
+        return { true, 0, "", storage[frameid] };
+    }
+};
+
+class YARP_dev_API yarp::dev::IFrameSource
+{
+    enum ConnectionType
+    {
+        DIRECT,
+        INVERSE,
+        UNDIRECT,
+        DISCONNECTED
+    };
+    ConnectionType getConnectionType(const std::string& target_frame, const std::string& source_frame, std::string* commonAncestor = nullptr);
+
+    robotology::Result<yarp::math::FrameTransform> getChainedTransform(const std::string& target_frame_id, const std::string& source_frame_id);
+
+    virtual bool frameExistsRaw(const std::string &frame_id);
+
+    FrameEditor frameContainer;
```

Comment:
I guess it should be possible to expose a `IFrameSource` interface from a class that **does not store** its frame information in a `FrameEditor`, so it could make sense to avoid inserting the `FrameEditor` in the virtual interface, right?

Reply:
mmm.. let me think.. actually the user calls the **already implemented** method of the virtual interface, which comunicate with the full-concrete layer via the intra-class protected interface using `FrameEditor` as a container. it does not make sense to remove the FrameEditor unless someone want to reimplement **all** of the interface method. i presume that to be a case less likely to happen respect to implement only the three protected method so i think it makes sense to not removing the FrameEditor. also because if you remove the frame editor you would force every device to implement all the method.
it's not the derived class that store its frame in the FrameEditor. It is the implemented side of the parent virtual class (IFrameSource) that use it to store the frames. The IFrameSource is not thinked  as pure interface though it can also be used as it was.

Reply:
Personal opinion: for consistency with the rest of the YARP interfaces, I would leave the IFrameSource as a pure virtual class, and move the default implementation that rely on the three protected methods to a separate class that inherits from `IFrameSource`, like a `IFrameSourceHelper` or a better name.